Mathlib/CategoryTheory/Sites/Coherent/RegularSheaves.lean

```diff
@@ -26,53 +26,140 @@ namespace CategoryTheory
   single_epi : ∃ (Y : C) (f : Y ⟶ X), R = Presieve.ofArrows (fun (_ : Unit) ↦ Y)
     (fun (_ : Unit) ↦ f) ∧ EffectiveEpi f
 
-namespace regularCoverage
+namespace regularTopology
+
+lemma equalizerCondition_w (P : Cᵒᵖ ⥤ D) {X B : C} {π : X ⟶ B} (c : PullbackCone π π) :
+    P.map π.op ≫ P.map c.fst.op = P.map π.op ≫ P.map c.snd.op := by
+  simp only [← Functor.map_comp, ← op_comp, c.condition]
 
 /--
-The map to the explicit equalizer used in the sheaf condition.
+A contravariant functor on `C` satisfies `EqualizerCondition` if it takes kernel pairs of effective
+epimorphisms to equalizer diagrams.
 -/
+def EqualizerCondition (P : Cᵒᵖ ⥤ D) : Prop :=
+  ∀ ⦃X B : C⦄ (π : X ⟶ B) [EffectiveEpi π] (c : PullbackCone π π) (_ : IsLimit c),
+    Nonempty (IsLimit (Fork.ofι (P.map π.op) (equalizerCondition_w P c)))
+
+/-- The equalizer condition is preserved by natural isomorphism. -/
+theorem equalizerCondition_of_natIso {P P' : Cᵒᵖ ⥤ D} (i : P ≅ P')
+    (hP : EqualizerCondition P) : EqualizerCondition P' := fun X B π _ c hc ↦
+  ⟨Fork.isLimitOfIsos _ (hP π c hc).some _ (i.app _) (i.app _) (i.app _)⟩
+
+/-- Precomposing with a pullback-preserving functor preserves the equalizer condition. -/
+theorem equalizerCondition_precomp_of_preservesPullback (P : Cᵒᵖ ⥤ D) (F : E ⥤ C)
+    [∀ {X B} (π : X ⟶ B) [EffectiveEpi π], PreservesLimit (cospan π π) F]
+    [F.PreservesEffectiveEpis] (hP : EqualizerCondition P) : EqualizerCondition (F.op ⋙ P) := by
+  intro X B π _ c hc
+  have h : P.map (F.map π).op = (F.op ⋙ P).map π.op := by simp
+  refine ⟨(IsLimit.equivIsoLimit (ForkOfι.ext ?_ _ h)) ?_⟩
+  · simp only [Functor.comp_map, op_map, Quiver.Hom.unop_op, ← map_comp, ← op_comp, c.condition]
+  · refine (hP (F.map π) (PullbackCone.mk (F.map c.fst) (F.map c.snd) ?_) ?_).some
+    · simp only [← map_comp, c.condition]
+    · exact (isLimitMapConePullbackConeEquiv F c.condition)
+        (isLimitOfPreserves F (hc.ofIsoLimit (PullbackCone.ext (Iso.refl _) (by simp) (by simp))))
+
+/-- The canonical map to the explicit equalizer. -/
 def MapToEqualizer (P : Cᵒᵖ ⥤ Type*) {W X B : C} (f : X ⟶ B)
     (g₁ g₂ : W ⟶ X) (w : g₁ ≫ f = g₂ ≫ f) :
     P.obj (op B) → { x : P.obj (op X) | P.map g₁.op x = P.map g₂.op x } := fun t ↦
   ⟨P.map f.op t, by simp only [Set.mem_setOf_eq, ← FunctorToTypes.map_comp_apply, ← op_comp, w]⟩
 
-/--
-The sheaf condition with respect to regular presieves, given the existence of the relavant pullback.
--/
-def EqualizerCondition (P : Cᵒᵖ ⥤ Type*) : Prop :=
+/-- The explicit equalizer condition. -/
+def EqualizerConditionExpl (P : Cᵒᵖ ⥤ Type*) : Prop :=
   ∀ (X B : C) (π : X ⟶ B) [EffectiveEpi π] [HasPullback π π], Function.Bijective
     (MapToEqualizer P π (pullback.fst (f := π) (g := π)) (pullback.snd (f := π) (g := π))
     pullback.condition)
 
+/-- The explicit equalizer condition is equivalent to the formal one. -/
+theorem equalizerCondition_iff_expl (P : Cᵒᵖ ⥤ Type*) :
+    EqualizerCondition P ↔ EqualizerConditionExpl P := by
```

Comment:
In terms of API, I am not sure we need to introduce `EqualizerConditionExpl`: I would rather see this as a theorem `EqualizerCondition.bijective_mapToEqualizer_pullback`, and the converse would be a constructor `EqualizerCondition.mk`.